pkg/apis/flowcontrol/validation/validation.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"math/big"
+
+	"k8s.io/apimachinery/pkg/api/validation"
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	apivalidation "k8s.io/kubernetes/pkg/apis/core/validation"
+	"k8s.io/kubernetes/pkg/apis/flowcontrol"
+)
+
+var ValidateFlowSchemaName = apimachineryvalidation.NameIsDNSSubdomain
+
+var ValidatePriorityLevelConfigurationName = apimachineryvalidation.NameIsDNSSubdomain
+
+var supportedDistinguisherMethods = sets.NewString(
+	string(flowcontrol.FlowDistinguisherMethodByNamespaceType),
+	string(flowcontrol.FlowDistinguisherMethodByUserType),
+)
+
+var supportedVerbs = sets.NewString(
+	flowcontrol.VerbAll,
+	"get",
+	"list",
+	"watch",
+	"create",
+	"update",
+	"delete",
+	"deletecollection",
+	"patch",
+)
+
+var supportedSubjectKinds = sets.NewString(
+	flowcontrol.ServiceAccountKind,
+	flowcontrol.UserKind,
+	flowcontrol.GroupKind,
+)
+
+const (
+	maxHashBits = 64
```

Comment:
We should set the upper limit below 64 bits to limit the unfairness of the coin flips.  I recommend setting the limit at 60 bits.

Reply:
`60` sounds tricky, why is it better than `64` do you think?
@mars1024

Reply:
If we have an array A0, A1, A2, ... AH, I think in extreme case, we have to calculate a limit value with AH=0, A0=(queueSize - 1), A1=(queueSize - 2), A2=(queueSize - 3), and if bits of this limit value is less or equal than 64, A0 - A(H-1) is valid for us.

Reply:
like this
```
const maxHashBits = 64

func ValidateParameters(queueSize, handSize uint32) bool {
	if handSize == 0 || queueSize == 0 || handSize > queueSize {
		return false
	}

	p := big.NewInt(0)
	for i := uint32(handSize); i > 0; i-- {
		p.Add(p.Mul(p, big.NewInt(int64(queueSize-i+1))), big.NewInt(int64(queueSize-i)))
	}

	return p.BitLen() <= maxHashBits
}
```
cc @yue9944882  @MikeSpreitzer

Reply:
After talking with @yue9944882 , I think we need a less time-consuming algorithm which allows a little calculation error.

Reply:
@mars1024 : you wrote an algorithm that I think was intended to calculate `FallingFactorial(NumQueues, HandSize) - 1`.  It gets simpler if you omit the `-1`!  But we can get even simpler.  In any scenario where limits are in question, I think we can expect that HandSize is much less than NumQueues.  So we can approximate FallingFactorial(NumQueues, HandSize) by `math.pow(NumQueues, HandSize)`.  In this case we can simply approximate the number of needed bits by `HandSize * math.Log2(NumQueues)`.

To illustrate why we want to not use all the hash bits, consider the example of NumQueues=512 and HandSize=7.  The last index is drawn from 506 possible values, but they do not have equal probabilities.  `2^64 / (512*511*510 * 509*508*507)` is 1054.  That is, when dealing the last card, the remaining hash value is in the range [0, 1053].  The last card will be the remainder of that remaining hash value divided by 506.  For 42 of the possible remainders the probability of that remainder appearing is 0.0059 and the other 464 each has a probability of 0.0040.  That is a pretty unfair coin flip.  It is pretty unfair because 506 is not a lot smaller than the number of remaining hash values.  It is only a little more than a factor of 2 smaller, so the last remainder probabilties have a 2:3 ratio.  If we limit the falling factorial to consuming less than 60 bits then the remainder probabilities will have a ratio of 16:17 or closer.

Reply:
@MikeSpreitzer  I've got your point, if the last dividend is one order of magnitude as the last divisor, the remainders may become uneven. So we need a proper bit length. I have a question about the ratio, I think 60 bits will make the ratio low to 32:33 instead of 16:17, so should we use 60bits or 61bits?

Reply:
updated to 60

Reply:
In the example I cited, the falling factorial takes just a little under 63 bits.  That made the ratio 2:3.  So another factor of 8 gets the ratio to 16:17.